### PR TITLE
refactor: long-standing cleanup tasks

### PR DESCRIPTION
**Pull request by DerekStrickland**
**NOTE TO REVIEWERS**: This is a PR around code organization and file/struct naming. No logic changes were made. You do not need to review implementation logic.

## Summary

This PR addresses several long standing cleanup tasks, primarily around code organization and naming. Because of the file moves and renaming, a very large number of files were affected.  To make it easier to review, the different refactoring tasks outlined below were done as separate commits but in the same PR. This should make reviewing each effort simpler for the reviewer. Separate PRs, would have made this effort take far too long amount of time. 

### Added

N/A

#### CLI

* Changed the `harbor-cli` binary name to `harbor`.
* Renamed the `sbom` command to `ingest` so that it is a verb that matches our domain language.
* Created module directories for each command (ingest, enrich, analyze).
* Split up large mod.rs files in each module directory so we have more focused code files.

#### Directory Structure

* Moved the `core` crate to the `sdk` folder.
* Moved the `platform` crate to the `sdk` folder.
* Moved the `mongodb` module under the `persistence` module in the `platform` crate.
* Renamed `book` folder to `docs`.
* Updated paths in the documentation that reference `core` or `platform`.
* Updated the workspace and crate-specific Cargo.toml files to reflect the changes.
* Refactored any code that is affected by the module structure changes.

#### Core

* Moved the `TaskProvider` implementations to a top level module called `tasks`.
* Renamed the various `provider.rs` files to their respective task names.
* Made sure task names align with commands.
* Refactored any code that is affected by the module structure changes.
* 
### Deprecated

* Using structs (e.g. EnrichmentProvider) to handle CLI commands. Those structs predated the use of `clap` which has 
  it's own CommandFactory functionality essentially. Functions were simpler and sufficient given that feature of `clap`.
* Using a `new_provider` function to instantiate `TaskProvider` implementations. The `TaskProviders` all have their own 
  `new` functions and this was straight forward to inline.

### Removed

* Deleted the no-longer used `client` crate.
* Deleted the no-longer used `pilot` crate.
* Deleted any functions and tests I could find that weren't actually called by anything anymore.

### Fixed

N/A

## How to test

Pull the branch and run `cargo test` to make sure it all compiles and the tests work.  Each commit is actually runnable and was tested along the way.

### CODE REVIEW

File: sdk/core/src/lib.rs [hunk 1]
```diff
@@ -66,12 +66,15 @@
 ///     }
 /// }
 /// ```
-
```
Comment (sbolel) [inline sdk/core/src/lib.rs:69]: Can removal of unnecessary newlines be enforced/fixed using a linter rule?

File: sdk/core/src/entities/teams/team.rs [hunk 2]
```diff
@@ -61,16 +61,4 @@ impl Team {
...
-    pub(crate) fn get_sbom_token(&self) -> Option<&str> {
-        let sbom_token = self
-            .tokens
-            .iter()
-            .filter(|t| t.enabled && t.expired().unwrap_or(true))
-            .map(|t| &*t.token)
-            .next();
-
-        sbom_token
-    }
```
Comment (sbolel) [inline sdk/core/src/entities/teams/team.rs:75]: Are we planning to replace this with checking for jwt tokens?

File: sdk/core/src/config.rs [hunk 3]
```diff
@@ -81,10 +80,7 @@ impl DocDbConfig {
                 "{}/?ssl=true&tlsCAFile=rds-combined-ca-bundle.pem&retryWrites=false",
                 connection_uri
             ),
-            false => format!(
-                "{}/?ssl=false&retryWrites=false",
-                connection_uri
-            )
+            false => format!("{}/?ssl=false&retryWrites=false", connection_uri),
```
Comment (sbolel) [inline sdk/core/src/config.rs:83]: Is this using a linter rule, if not can it be?

File: sdk/core/src/services/vulnerabilities/mod.rs [hunk 4]
```diff
@@ -127,11 +121,7 @@ impl StorageProvider for S3StorageProvider {
         let s3_store = s3::Store::new_from_env().await?;
         let bucket_name = config::harbor_bucket()?;
 
-        let object_key = format!(
-            "vulnerabilities-{}-{}",
-            provider,
-            make_s3_key_safe(purl)?
-        );
+        let object_key = format!("vulnerabilities-{}-{}", provider, make_s3_key_safe(purl)?);
```
Comment (sbolel) [inline sdk/core/src/services/vulnerabilities/mod.rs:124]: Is this enforced with a linter rule? If not, can it be?

File: sdk/platform/src/persistence/mongodb/analytics.rs [hunk 5]
```diff
@@ -21,11 +19,10 @@ pub struct Pipeline {
     store: Arc<Store>,
     /// This variable holds all of the Stages to be executed
     /// in the pipeline.
-    stages: Mutex<Vec<Stage>>
+    stages: Mutex<Vec<Stage>>,
```
Comment (sbolel) [inline sdk/platform/src/persistence/mongodb/analytics.rs:22]: Is this enforced with a linter rule? If not, can it be?

File: sdk/platform/src/persistence/mongodb/analytics.rs [hunk 6]
```diff
@@ -46,13 +43,10 @@ impl Pipeline {
...
-
         if self.stages.lock().unwrap().len() == 0 {
-            return Err(
-                Error::Mongo(
-                    String::from("The Aggregation Pipeline is empty")
-                )
-            )
+            return Err(Error::Mongo(String::from(
+                "The Aggregation Pipeline is empty",
+            )));
```
Comment (sbolel) [inline sdk/platform/src/persistence/mongodb/analytics.rs:49]: Same comment here - Is this enforced with a linter rule? If not, can it be?

File: sdk/platform/src/persistence/mongodb/analytics.rs [hunk 7]
```diff
@@ -63,7 +57,8 @@ impl Pipeline {
         let options = AggregateOptions::builder().build();
 
         // Map the stages over to Documents
-        let doc_pipeline = self.stages
+        let doc_pipeline = self
+            .stages
```
Comment (sbolel) [inline sdk/platform/src/persistence/mongodb/analytics.rs:61]: Same comment here - Is this enforced with a linter rule? If not, can it be?

File: sdk/platform/src/persistence/mongodb/analytics.rs [hunk 8]
```diff
@@ -76,34 +71,27 @@ impl Pipeline {
...
                 self.clear();
 
-                Err(
-                    Error::Mongo(
-                        String::from("No result from DocumentDB Aggregate")
-                    )
-                )
+                Err(Error::Mongo(String::from(
+                    "No result from DocumentDB Aggregate",
+                )))
```
Comment (sbolel) [inline sdk/platform/src/persistence/mongodb/analytics.rs:94]: and same here - Is this enforced with a linter rule? If not, can it be?

File: cli/src/commands/enrich/mod.rs [hunk 9]
```diff
@@ -19,11 +18,11 @@ pub enum EnrichmentProviderKind {
...
 }
 
-/// The Enrich Command handler.
+/// The CommandFactory function for the `enrich` command.
 pub async fn execute(args: &EnrichArgs) -> Result<(), Error> {
     match args.provider {
-        EnrichmentProviderKind::Epss => EpssProvider::execute(args).await,
-        EnrichmentProviderKind::Snyk => SnykProvider::execute(args).await,
+        EnrichmentProviderKind::Epss => epss::execute(args).await,
+        EnrichmentProviderKind::Snyk => snyk::execute(args).await,
```
Comment (sbolel) [inline cli/src/commands/enrich/mod.rs:25]: cleaner 👍

File: cli/src/commands/analyze/sbom_detail.rs [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+            let mut task: Task = Task::new(TaskKind::Analytics(AnalyticProviderKind::SbomDetail))
+                .map_err(|e| Error::Analyze(e.to_string()))?;
+
+            provider
+                .execute(&mut task)
+                .await
+                .map_err(|e| Error::Analyze(e.to_string()))
+        }
+        Some(_a) => Err(Error::Analyze(String::from(
+            "other analysis not yet implemented",
```
Comment (sbolel) [inline cli/src/commands/analyze/sbom_detail.rs:54]: Is the state of this documented anywhere?

File: cli/src/commands/enrich/snyk.rs [hunk 11]
```diff
@@ -26,98 +26,74 @@ pub struct SnykArgs {
...
-            Some(_a) => Err(Error::Enrich(
-                "individual project not yet implemented".to_string(),
-            )),
+            provider
+                .execute(&mut task)
+                .await
+                .map_err(|e| Error::Enrich(e.to_string()))
         }
+        Some(_a) => Err(Error::Enrich(
+            "individual project not yet implemented".to_string(),
```
Comment (sbolel) [inline cli/src/commands/enrich/snyk.rs:74]: Is the state of this documented anywhere?

File: cli/src/commands/ingest/mod.rs [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+    type Err = ();
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let value = s.to_lowercase();
+        let value = value.as_str();
+        match value {
+            "filesystem" | "f" => Ok(IngestionProviderKind::FileSystem),
+            "github" | "gh" | "g" => Ok(IngestionProviderKind::GitHub),
+            "snyk" | "s" => Ok(IngestionProviderKind::Snyk),
+            _ => Err(()),
```
Comment (sbolel) [inline cli/src/commands/ingest/mod.rs:88]: This is a pretty cool syntax for building commands with aliases. Nice.

File: cli/src/commands/ingest/snyk.rs [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+
+            provider
+                .execute(&mut task)
+                .await
+                .map_err(|e| Error::Sbom(e.to_string()))
+        }
+        Some(args) => {
+            let (_, _) = (&args.org_id, &args.project_id);
+            Err(Error::Sbom(
+                "individual projects not yet implemented".to_string(),
```
Comment (sbolel) [inline cli/src/commands/ingest/snyk.rs:73]: Is the state of this documented anywhere?

**Review by sbolel — APPROVED** [review]
🤩 Looks great, and it looks like quite a big lift -- nice work! I really like how you divide your PRs into thematically-grouped commits. 

I left some minor comments and questions on syntax, tooling, and debug statements (apologies for the repetitive ones).

Test results:
- `cargo build` ✅ 
- `cargo test`  🛑 

The failing test is below:

```
running 1 test
test can_init_auth ... FAILED

failures:

---- can_init_auth stdout ----
Error: Migration("error in mongo provider: Error { kind: ServerSelection { message: \"Server selection timeout: No available servers. Topology: { Type: Unknown, Servers: [ { Address: mongo:27017, Type: Unknown, Error: Kind: I/O error: failed to lookup address information: nodename nor servname provided, or not known, labels: {} } ] }\" }, labels: {}, wire_version: None, source: None }")


failures:
    can_init_auth

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 30.02s

error: test failed, to rerun pass `-p harbcore --test migrations`
```

I assume I'm missing the step of starting up mongo here?

**Comment by DerekStrickland** [comment]
> 🤩 Looks great, and it looks like quite a big lift -- nice work! I really like how you divide your PRs into thematically-grouped commits.
> 
> I left some minor comments and questions on syntax, tooling, and debug statements (apologies for the repetitive ones).
> 
> Test results:
> 
> * `cargo build` ✅
> * `cargo test`  🛑
> 
> The failing test is below:
> 
> ```
> running 1 test
> test can_init_auth ... FAILED
> 
> failures:
> 
> ---- can_init_auth stdout ----
> Error: Migration("error in mongo provider: Error { kind: ServerSelection { message: \"Server selection timeout: No available servers. Topology: { Type: Unknown, Servers: [ { Address: mongo:27017, Type: Unknown, Error: Kind: I/O error: failed to lookup address information: nodename nor servname provided, or not known, labels: {} } ] }\" }, labels: {}, wire_version: None, source: None }")
> 
> 
> failures:
>     can_init_auth
> 
> test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 30.02s
> 
> error: test failed, to rerun pass `-p harbcore --test migrations`
> ```
> 
> I assume I'm missing the step of starting up mongo here?

Yeah sorry.  That's in the README. I included the same text on several commits in the past, but then eventually stopped repeating myself and assumed everyone knew how to run the devenv by now.  I later learned that wasn't necessarily true. In the future I'll make sure to point everyone to the README first to run the devenv,.

File: sdk/core/src/lib.rs (hunk 1 above)
Comment (DerekStrickland) [reply sdk/core/src/lib.rs:69]: It probably is, unfortunately, we have a problem with our current pre-commit hooks and I have to disable them when splitting up large commits. The way they work right now, you have to add all current pending files to the current commit, so you can't break large commits apart. You have to plan it all in advance which is pretty challenging to anticipate. Maybe you can help me figure that out.

File: sdk/core/src/entities/teams/team.rs (hunk 2 above)
Comment (DerekStrickland) [reply sdk/core/src/entities/teams/team.rs:75]: I believe we intend to have both long term. JWT for generally Authentication/Authorization and tokens as an additional factor specifically for uploads.

File: cli/src/commands/analyze/sbom_detail.rs (hunk 10 above)
Comment (DerekStrickland) [reply cli/src/commands/analyze/sbom_detail.rs:54]: Only in the rough draft stuff we're working through together with Eric.

File: cli/src/commands/enrich/snyk.rs (hunk 11 above)
Comment (DerekStrickland) [reply cli/src/commands/enrich/snyk.rs:74]: No, but stay tuned. 😉

File: cli/src/commands/ingest/mod.rs (hunk 12 above)
Comment (DerekStrickland) [reply cli/src/commands/ingest/mod.rs:88]: agreed 😁

File: cli/src/commands/ingest/snyk.rs (hunk 13 above)
Comment (DerekStrickland) [reply cli/src/commands/ingest/snyk.rs:73]: No. Where were you imagining that to be documented?  Backlog or somewhere else?

File: sdk/core/src/config.rs (hunk 3 above)
Comment (DerekStrickland) [reply sdk/core/src/config.rs:83]: See previous comment about linter and splitting large commits.

File: sdk/platform/src/persistence/mongodb/analytics.rs (hunk 5 above)
Comment (DerekStrickland) [reply sdk/platform/src/persistence/mongodb/analytics.rs:22]: See previous comment about linter and splitting large commits.

File: sdk/platform/src/persistence/mongodb/analytics.rs (hunk 7 above)
Comment (DerekStrickland) [reply sdk/platform/src/persistence/mongodb/analytics.rs:61]: See previous comment about linter and splitting large commits.

File: sdk/platform/src/persistence/mongodb/analytics.rs (hunk 8 above)
Comment (DerekStrickland) [reply sdk/platform/src/persistence/mongodb/analytics.rs:94]: See previous comment about linter and splitting large commits.

**Comment by sbolel** [comment]
> Yeah sorry. That's in the README. I included the same text on several commits in the past, but then eventually stopped repeating myself and assumed everyone knew how to run the devenv by now. I later learned that wasn't necessarily true. In the future I'll make sure to point everyone to the README first to run the devenv,.

Ah! Sounds good -- I'll check the readme and complete the necessary steps and then come back to this!

File: cli/src/commands/ingest/snyk.rs (hunk 13 above)
Comment (sbolel) [reply cli/src/commands/ingest/snyk.rs:73]: I wasn't sure if there was a standard for it. I'm used to seeing ``// TODO: ...` comments in JS which makes it easier to track down. It's not an issue -- it was more for my own understanding 👍

File: sdk/core/src/lib.rs (hunk 1 above)
Comment (sbolel) [reply sdk/core/src/lib.rs:69]: Ah, that sounds like an ordeal. Okay, let's take note of that and come back to it. I think @jonshattuck was also having some trouble with the pre-commit hooks messing with files 🤔

**Review by sbolel — DISMISSED** [review]
All tests passed ✅ :shipit:

File: sdk/core/src/lib.rs (hunk 1 above)
Comment (sbolel) [reply sdk/core/src/lib.rs:69]: I added a task card in Coda to track this work 👍

File: sdk/core/src/entities/analytics/mod.rs [hunk 14]
```diff
@@ -14,4 +14,4 @@ impl Display for AnalyticProviderKind {
             AnalyticProviderKind::SbomDetail => write!(f, "sbom-detail"),
         }
     }
-}
```
Comment (sbolel) [inline sdk/core/src/entities/analytics/mod.rs:17]: This makes me happy.

**Review by sbolel — APPROVED** [review]
:shipit:

File: sdk/core/src/entities/analytics/mod.rs [hunk 15]
```diff
@@ -14,4 +14,4 @@ impl Display for AnalyticProviderKind {
...
\ No newline at end of file
+}
```
Comment (sbolel) [inline sdk/core/src/entities/analytics/mod.rs:17]: Doin' the Lord's work. I LOVE EOF NEWLINES.

**Pull request merged**